lib/commands/streamscreen.js

```diff
@@ -0,0 +1,408 @@
+ * @property {?number} quality [70] - The quality value for the streamed JPEG images.
+ * This number should be in range [1, 100], where 100 is the best quality.
+ * @property {?boolean} considerRotation [false] - If set to `true` then GStreamer pipeline will
+ * increase the dimensions of the resulting images to properly fit images in both landscape and
+ * portrait orientations. Set it to `true` if the device rotation is not going to be the same during the
+ * broadcasting session.
+ * @property {?boolean} logPipelineDetails [false] - Whether to log GStreamer pipeline events into
+ * the standard log output. Might be useful for debugging purposes.
+ */
+
+/**
+ * Starts device screen broadcast by creating MJPEG server.
+ * Multiple calls to this method have no effect unless the previous streaming
+ * session is stopped.
+ *
+ * @param {?StartScreenStreamingOptions} options - The available options.
+ * @throws {Error} If screen streaming has failed to start or is not supported on the host system.
+ */
+commands.mobileStartScreenStreaming = async function mobileStartScreenStreaming (options = {}) {
+  const {
+    width,
+    height,
+    bitRate,
+    host = DEFAULT_HOST,
+    port = DEFAULT_PORT,
+    tcpPort = DEFAULT_PORT + 1,
+    quality = DEFAULT_QUALITY,
+    considerRotation = false,
+    logPipelineDetails = false,
+  } = options;
+
+  if (_.isUndefined(this._screenStreamingProps)) {
+    await verifyStreamingRequirements(this.adb);
+  }
+  if (!_.isEmpty(this._screenStreamingProps)) {
+    log.info(`The screen streaming session is already running. ` +
+      `Stop it first in order to start a new one.`);
+    return;
+  }
+  if ((await checkPortStatus(port, host)) === 'open') {
+    log.info(`The port #${port} at ${host} is busy. ` +
+      `Assuming the screen streaming is already running`);
+    return;
+  }
+  if ((await checkPortStatus(tcpPort, TCP_HOST)) === 'open') {
+    log.errorAndThrow(`The port #${tcpPort} at ${TCP_HOST} is busy. ` +
+      `Make sure there are no leftovers from previous sessions.`);
+  }
+  this._screenStreamingProps = null;
+
+  const deviceInfo = await getDeviceInfo(this.adb);
+  const deviceStreamingProc = await initDeviceStreamingProc(this.adb, deviceInfo, {
+    width,
+    height,
+    bitRate,
+  });
+  let gstreamerPipeline;
+  try {
+    gstreamerPipeline = await initGstreamerPipeline(deviceStreamingProc, deviceInfo, {
+      width,
+      height,
+      quality,
+      tcpPort,
+      considerRotation,
+      logPipelineDetails,
+    });
+  } catch (e) {
+    if (deviceStreamingProc.kill(0)) {
+      deviceStreamingProc.kill();
+    }
+    throw e;
+  }
+
+  let mjpegSocket;
+  let mjpegServer;
+  try {
+    await new B((resolve, reject) => {
+      mjpegSocket = net.createConnection(tcpPort, TCP_HOST, () => {
+        log.info(`Successfully connected to MJPEG stream at tcp://${TCP_HOST}:${tcpPort}`);
+        mjpegServer = http.createServer((req, res) => {
```

Comment:
is there no existing mjpeg server package we can use? or, should we extract this into its own utility?

Reply:
I prefer to implement the stuff on my own if it is tiny rather than adding more dependencies. Nevertheless feel free to change that if you think the existing package would perform better or contains something the current solution does not have